resdk/data_upload/samplesheet.py

```diff
@@ -0,0 +1,518 @@
+        """Check is the provided path exists."""
+        if not os.path.isfile(path):
+            raise OSError(
+                "The provided annotation file '{}' "
+                "does not exist.".format(path)
+            )
+
+    def _get_spreadsheet_extension(self, path):
+        """Find spreadsheet file extension."""
+        return os.path.splitext(path)[1]
+
+    def _read_xls(self, path):
+        """Read Excel spreadsheet annotation file."""
+        workbook = load_workbook(path)
+        worksheet = workbook.active
+        rows = worksheet.rows
+        header = [cell.value for cell in next(rows)]
+        return [self._parse_row(header, row) for row in rows]
+
+    def _parse_row(self, header, row):
+        """Convert spreadsheet row into sample entry."""
+        return {head: self._parse_cell(cell) for head, cell in zip(header, row)}
+
+    def _parse_cell(self, cell):
+        """Interpret spreadsheet cell."""
+        if isinstance(cell.value, float):
+            return str(cell.value)
+        elif cell.value in EMPTY:
+            return ''
+        else:
+            return cell.value
+
+    def _read_text_file(self, path):
+        """Read simple spreadsheet annotation file."""
+        with open(path, 'rb') as sample_sheet:
+            return list(csv.DictReader(sample_sheet, delimiter='\t'))
+
+    def _populate_entries(self, path):
+        """Check the format of annotation file and assign read function."""
+        if self._get_spreadsheet_extension(path) in ['.xls', '.xlsx', '.xlsm']:
+            return self._read_xls(path)
+        elif self._get_spreadsheet_extension(path) in ['.txt', '.tab', '.tsv']:
+            return self._read_text_file(path)
+        else:
+            raise TypeError(
+                "Annotation spreadsheet extension '{}' not recognised. Options"
+                " are: '.xls', '.xlsx', '.xlsm', '.txt', '.tab', "
+                "'.tsv'.".format(self._get_spreadsheet_extension(path))
+            )
+
+    def _create_all_samples(self, entries):
+        """Create a sample from each samplesheet entry."""
+        valid_samples = OrderedDict()
+        invalid_names = set()
+        for entry in entries:
+            valid_samples, invalid_names = (
+                self._create_sample(entry, valid_samples, invalid_names)
+            )
+        return valid_samples, invalid_names
+
+    def _create_sample(self, entry, valid_samples, invalid_names):
+        """Create a sample from a samplesheet entry, if unique."""
+        name = entry['SAMPLE_NAME']
+        if name in invalid_names or name in valid_samples:
+            logger.error(
+                "The sample name '%s' is duplicated. Please use unique "
+                "sample names.",
+                name
+            )
+            invalid_names.add(name)
+            valid_samples.pop(name, None)
+        else:
+            try:
+                valid_samples[name] = Sample(entry)
+            except ValueError as ex:
```

Comment:
I don't think we want to catch the `KeyError`, as it would indicated that the column headers have changed, which means the entire annotation spreadsheet is problematic.